### PR TITLE
Refactor SparseApply[Proximal]Adagrad into classes

### PR DESCRIPTION
**Pull request by benbarsdell**
- This is a NFC in preparation for adding GPU implementations.
- It also combines the SparseApplyAdagrad[V2] kernels into a single implementation distinguished by a has_epsilon template parameter.

cc @nluehr

### CODE REVIEW

File: tensorflow/core/kernels/training_ops.h [hunk 1]
```diff
@@ -92,6 +92,17 @@ struct ApplyAdagradDA {
                   typename TTypes<T>::ConstFlat grad);
 };
 
+template <typename Device, typename T, typename Tindex, bool has_epsilon>
+struct SparseApplyAdagrad {
+  Status operator()(const Device& d, typename TTypes<T>::Matrix var,
+                    typename TTypes<T>::Matrix accum,
+                    typename TTypes<T>::ConstScalar lr,
+                    typename TTypes<T>::ConstScalar epsilon,
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.h:100]: Please document what is the semantic of `epsilon` if `has_epsilon` is false (it is ignored right?).

File: tensorflow/core/kernels/training_ops.cc [hunk 2]
```diff
@@ -162,6 +162,90 @@ struct ApplyAdagradV2<CPUDevice, T> {
...
+              v -= g.constant(lr_scalar) * g /
+                   (a.sqrt() + a.constant(epsilon()));
+            } else {
+              v -= g.constant(lr_scalar) * g * a.rsqrt();
+            }
+          }
+        };
+
+        d.parallelFor(N, cost, shard);
+
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:213]: Stray newline?

File: tensorflow/core/kernels/training_ops.cc [hunk 3]
```diff
@@ -162,6 +162,90 @@ struct ApplyAdagradV2<CPUDevice, T> {
...
+                    typename TTypes<T>::Matrix accum,
+                    typename TTypes<T>::ConstScalar lr,
+                    typename TTypes<T>::ConstScalar epsilon,
+                    typename TTypes<T>::ConstMatrix grad,
+                    typename TTypes<Tindex>::ConstVec indices, int64 inner_dim,
+                    bool update_slots) {
+    const Tindex N = static_cast<Tindex>(indices.dimension(0));
+    const Tindex first_dim_size = static_cast<Tindex>(var.dimension(0));
+    const T lr_scalar = lr();
+    if (N > 0) {
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:177]: Please use early exit to reduce indentation.

File: tensorflow/core/kernels/training_ops.cc [hunk 4]
```diff
@@ -190,6 +274,79 @@ struct ApplyProximalAdagrad<CPUDevice, T> {
...
+                    typename TTypes<T>::ConstScalar l2,
+                    typename TTypes<T>::ConstMatrix grad,
+                    typename TTypes<Tindex>::ConstVec indices,
+                    int64 inner_dim) {
+    const Tindex N = static_cast<Tindex>(indices.dimension(0));
+    const Tindex first_dim_size = static_cast<Tindex>(var.dimension(0));
+    const T lr_scalar = lr();
+    const T l1_scalar = l1();
+    const T l2_scalar = l2();
+    if (N > 0) {
```
Comment (sanjoy) [inline tensorflow/core/kernels/training_ops.cc:292]: Please use early exit to reduce indentation.

**Review by sanjoy — CHANGES_REQUESTED** [review]
Only minor comments (I'm assuming this is just moving code around).

File: tensorflow/core/kernels/training_ops.h (hunk 1 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.h:101]: Done.

File: tensorflow/core/kernels/training_ops.cc (hunk 2 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc:221]: Fixed.

File: tensorflow/core/kernels/training_ops.cc (hunk 3 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc]: Done.

File: tensorflow/core/kernels/training_ops.cc (hunk 4 above)
Comment (benbarsdell) [reply tensorflow/core/kernels/training_ops.cc]: Done.

**Comment by irvineoy** [comment]
Sorry for commenting on this PR. I implemented the GPU kernel of ResourceSparseApplyAdadelta which is still waiting for review (#44347). And I found Adagrad optimizer will also encounter the same problem when dealing with sparse tensor on GPU. So I wonder do you consider implementing the GPU kernel of ResourceSparseApplyAdagrad as a good way to solve it and if this a good part to contribute? Thank you very much. @sanjoy @benbarsdell

**Comment by benbarsdell** [comment]
@irvineoy thanks for your interest. I have implemented a GPU version of the ops in this PR and will be submitting it in a follow-up PR as soon as this one is merged.

**Review by sanjoy — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**